### ISSUE TITLE
[publishing] Add support for the new maven central publisher API

### ISSUE THREAD

**Issue by martinbonnin** [gradle/gradle#28120]
### Expected Behavior

Sonatype is changing how artifacts are uploaded with new accounts. See https://central.sonatype.org/register/central-portal/ and https://central.sonatype.org/publish/publish-portal-api/

The new API uploads a single zip file containing a full publication in one go and as a result the current method doesn't work anymore. Would be nice to have built-in support for the new API in the `maven-publish` plugin.

### Context

https://github.com/vanniktech/gradle-maven-publish-plugin/issues/722
[KotlinLang slack discussion](https://slack-chats.kotlinlang.org/t/16407082/which-gradle-plugin-should-i-use-to-publish-kmp-project-arti#044659f2-81f1-4ea6-b899-2c35b3cc06c2)

**Comment by ljacomet** [comment]
This is something that we are discussing internally and probably will discuss with Sonatype.

An official plugin from Sonatype would be the best option since they own the protocol which is no longer based on Maven `deploy`.

If an official Gradle plugin adds that support, it would also make sense to have it _outside_ of the Gradle core plugins, since Sonatype could change their protocol anytime.

**Comment by martinbonnin** [comment]
Not sure if this is still on the table for Sonatype but keeping a similar API would be quite handy:

* no need to change every library using Gradle/Maven out there
* could parallelize the uploads and/or upload from different machines (for KMP for an example)

I get that the current implicit staging repos and the split repos are an issue but sounds like this could be fixed transparently for end users by wrapping the file uploads in creation/close API calls? 

While I understand the rationale for a new plugin, adding yet another publishing plugin is going to add more complexity to an already complex landscape.

Anyways, I understand this is ultimately not in Gradle hands but these were my 2 cents! Thanks for reading!

**Comment by asarkar** [comment]
@ljacomet 
> An official plugin from Sonatype would be the best option since they own the protocol

Sonatype has a [page](https://central.sonatype.org/publish/publish-portal-gradle/) listing dozens of Gradle plugins that use the Central Publishing Portal. There's no indication that they ever intend to create a plugin of their own. As for the protocol, if all these third-party plugins could implement it, why can't Gradle?

**Comment by hfhbd** [comment]
@asarkar 

> There's no indication that they ever intend to create a plugin of their own.

[It is](https://central.sonatype.org/publish/publish-portal-gradle/): `so we wanted to communicate that Gradle support is on our roadmap.`

**Comment by asarkar** [comment]
> so we wanted to communicate that Gradle support is on our roadmap

I saw that. Actual roadmaps have milestones and release dates, they have published none. That statement sounds no nothing more than an apology to the backlash for breaking the now legacy API. The Central API is now almost a year old.

**Comment by osiegmar** [comment]
Just received a reminder email from Sonatype that states:

> As you may have heard, OSSRH is reaching end of life on June 30, 2025. OSSRH users need to migrate their namespaces to the Central Portal as soon as possible.
> 
> Instructions for self migration are located here: [https://central.sonatype.org/faq/what-is-different-between-central-portal-and-legacy-ossrh/#self-service-migration](https://u21795309.ct.sendgrid.net/ls/click?upn=u001.cdbupahfLeeV0JrmnGgqhiqfh8Qc-2Bz8uDGMwn-2BLd344y41cU9pTHtj0CTURHePc0V2v32L21bp-2BAj2eo9txd-2B1-2BZCCn701Wxm5U0hO7l-2FeGczGHqDzDH-2FXq-2BdD-2FrcS066VxP62gE5WrdgqFu609C6WQie9hRwNsz-2BxogKUh1HGM-3D5aFt_6vELXDU9KZ1iFJaR21H661eBoiirHxoSVc-2BWnrMB5HzJ0RJX2-2FH9-2FtES9SOPYcWSdUXThfuEZFKeJRSV7z0qGunWFl8rIjoFoN-2BbEE-2Bq-2Bh5FJ-2BWqJopK2BZcgd35d04NkFtWoc87r078QcuUEcX7y3XzkL2BId2-2FTfEINtBv-2BEfW1m1EGX-2Bf-2BDs9Oj2CQwqJZ3WTYlTAoQuJLpv3Mm3ZaZeQ-2BZavzzV9wzTgFAaHIJ-2FunAS7mh66-2Fdgtrs99vM9NYbh3QUQbPhcMu3ZyK-2BKuKZEp-2BesI-2F9m10m512que47Aq0aXF510UyAY4VCgYWr6SImmCBi2o4FE88FJHLOKQwQuJEQCvXDQqkeWy461yZQKYLLiXSeXw3Hau3JSinad5KCCah7VdaDaExMHkA-2BMKzx-2FncEJc2xoO52VIGAyCuPLid2nyHo2Kc0Eac9MHbELTFN6q0zXCeZwMml5-2BZl30zA4TCEfpt3xyU1vtWO-2FCXvKUqerwkm3VVcG-2FHLDgJkFcwgVFzJqVTnKTZiYWy8VVjZt2e7jrbZrnUCjTD-2FyWvkp3xHeLtWBc5Yh5aKotKQA3zDSQ-2B8SEkAa87XgZtRwsCbsYxyJzpjVuJUK-2F8y-2F6qPzjHZkAJC5KPq4PWJGDcCVSAH58v8JiGUMD8uiXOwe0-2F1-2FRe-2FKD84KHNh81W0GAtoenQ0fC66-2B-2FKX8yVBagGHYYabd0QNHBUY-2F-2FtD8c-2BdHCYrjkNbOWhJk3yyX9UFwnnnLaui4qri-2BHNpTZxgJDYJRoBfSZpsOxLnxz0xI5lkRttBmL4DzS-2FuAh3GOZG6ZVmHIhxqhfz2NOOFydyQ3w4ZrgyDL84dFRKcEAJP8gTDK4np2-2BDpyf-2F70soo0nzgzFyVMULuLNALeVzeOLkvDPd2Ui2jsbPkrxLNFj6Zw-2BECMKYA-3D-3D)
> 
> To make the transition smoother we will be automatically migrating publishers that have not used oss.sonatype.org or s01.sonatype.org to publish artifacts in some time starting with the oldest and working our way forward. To avoid disruption to your publishing processes we strongly encourage migrating before the June 30, 2025 deadline.

Any update on this issue or advice on how to move forward?

**Comment by hfhbd** [comment]
@osiegmar there is a list of community plugins supporting the new protocol: https://central.sonatype.org/publish/publish-portal-gradle/

**Comment by osiegmar** [comment]
@hfhbd thanks – I’m aware of that list, but I was hoping for a final decision on this issue since it remains open. Sonatype’s statement about "having it on their roadmap" has yet to yield any results.

Giving access to my GPG key to a third-party plugin (of uncertain origin and quality) requires additional effort for code analysis and other precautions. Only a few plugins listed at https://central.sonatype.org/publish/publish-portal-gradle/ have shown recent activity or an established or growing user base.

If Sonatype fails to deliver and Gradle considers this beyond their scope, an official recommendation in Gradle’s documentation would be appreciated.

**Comment by jlink** [comment]
> In fact, there are so many publishing plugins because the api is quite simple, [you can easily write your plugin](https://github.com/hfhbd/mavencentral/tree/main/gradle-plugin/src/main/kotlin).

Still, it requires me to invest time into something that has no perceivable benefit for users. That might be the straw that pushes me to no longer maintain some of the OSS libraries for which I feel responsible.

**Comment by jlink** [comment]
Moreover, Sonatype is not a random org that invents a protocol that may or may not survive, but they are the most important column in the Maven-based repositories universe. You need not like that but it's something to consider for the decision to (not) support a protocol.

**Comment by hfhbd** [comment]
> Still, it requires me to invest time into something that has no perceivable benefit for users.

You can use any 3rd party plugins or the upcoming 1st party plugin.

> that invents a protocol that may or may not survive,

If so, why there is a new protocol and this issue at all? :)
It is absolute okay to change protocols, but I don't think it should be coupled with Gradles lifecycle.

**Comment by jlink** [comment]
> > Still, it requires me to invest time into something that has no perceivable benefit for users.
> 
> You can use any 3rd party plugins or the upcoming 1st party plugin.

Using, configuring and testing a 3rd party plugin is work - at least for me. Work that I do not like at all, using up time that could be spent elsewhere if the new protocol was "just" supported by Gradle's publishing plugin.

**Comment by asarkar** [comment]
As I've mentioned before, I also support the idea of Gradle providing official support for the Sonatype Central API. However, it's worth noting that this issue hasn't gained much traction among the 3,000+ open issues.

https://github.com/gradle/gradle/issues?q=is%3Aissue%20state%3Aopen%20sort%3Areactions-%2B1-desc

FWIW, I find the "Sonatype API might change" argument unconvincing, given that Gradle already supports third-party libraries like PMD through plugins.

**Comment by martinbonnin** [comment]
As the original poster here and maintainer of one of the [third party plugins](https://github.com/GradleUp/nmcp/), I agree with [@hfhbd comment above](https://github.com/gradle/gradle/issues/28120#issuecomment-2865261379) and am now team "this doesn't belong in Gradle". 

The Gradle feature set is huge already and I'd like it better if this kind of functionality could be decoupled so that Gradle can  focus on the core of the build system, leaving out other parts to the different third parties. Sonatype is the best positioned to provide this. 

Where it gets interesting is how to make different plugins communicate. My overall understanding today is:
1. Plugins like Java or Kotlin create `SoftwareComponents`.
1. Users create `MavenPublications` from `SoftwareComponents` in their build scripts.
1. The `maven-publish` uploads the `MavenPublication` to OSSRH or another "old" maven host.

Having Gradle define a common lingua franca how to model files is interesting because it means Java and Kotlin (and Scala and Groovy, and maybe C?) can speak the same language and you don't need a different publishing plugin for each ecosystem.

**Comment by asarkar** [comment]
After reviewing the plugins listed on the Sonatype page, it seems that _JReleaser_ and _vanniktech/gradle-maven-publish-plugin_ are the two most popular and actively maintained options. However, I encountered a few challenges with JReleaser. It appeared to require certain properties to be set by the user that’s weren’t very evident, leading to NPEs when those were missing. Additionally, I struggled to get it to sign artifacts correctly. As a result, I ultimately opted for the vanniktech plugin, which proved more befitting for my use case.

**Comment by Vampire** [comment]
Did anyone (especially the people that don't want to use 3rd party plugins) read the big info box right in the beginning of that page that lists the community plugins?

You can continue use the "old" community plugins or also `maven-publish`.
That info box links to https://central.sonatype.org/publish/publish-portal-ossrh-staging-api/ which describes that they implemented a bridge from the "old way" to the "new portal", so that you just have to use a new URL and generate a new user token on the new system.

---

tl;dr: You can continue using `maven-publish` or any plugin supporting the "old way"!

**Comment by Vampire** [comment]
Why are you thumbs-downing my comment @madhead, is something wrong in my comment?

**Comment by madhead** [comment]
@Vampire, please, don't treat it personal, because it's not.

This issue is specifically about the new API support. Recommending users to use the old API is not a proper solution IMHO.

Even the page you've linked says:

> We hope that over time plugins will adopt the [Portal API](https://central.sonatype.org/publish/publish-portal-api/) rather than rely on this service.

So, supporting the new API is inevitable.

**Comment by Vampire** [comment]
I don't take it personal, just wanted to know whether you see some factual error or just don't like it. :-)
I'm not saying the new API should not be supported.
But like others I'm not sure whether Gradle should support this non-standard way to publish.
But that's up to the Gradle folks to decide.

But in the last couple of days there was a heated debate in this issue, mainly because some people do not want to use a 3rd-party plugin for publishing, there is neither built-in support in Gradle yet, nor an official plugin by Sonatype yet, **and they assumed that they cannot use the classical way**. And this last assumption is just wrong, that's all I said to help those people that want to wait for 1st party support by either Sonatype or Gradle, so that they can for the time being - call it a work-around if you like - use the old mechanisms by just using a new URL. :-)

Also, it is not necessarily "inevitable", they did not state yet anywhere that they will stop maintaining that service, just that they "hope" people will use the new API. If 90% of the ecosystem uses that service instead of the new API, that could also be a signal to them that they might understand or not. :-D 🤷‍♂️

But let's not discuss this any further here, I think enough was said that pings all watchers. :-)
